File: lib/java/src/main/java/uapi/internal/generation/GenerateRandomUnion.java
package uapi.internal.generation;

import static uapi.internal.generation.GenerateRandomStruct.generateRandomStruct;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import uapi.internal.types.UStruct;
import uapi.internal.types.UTypeDeclaration;

public class GenerateRandomUnion {
    public static Object generateRandomUnion(Object blueprintValue,
            boolean useBlueprintValue, Map<String, UStruct> unionTagsReference,
            GenerateContext ctx) {
        if (useBlueprintValue) {
            final var startingUnion = (Map<String, Object>) blueprintValue;
            final var entry = startingUnion.entrySet().stream().findAny().get();
            final var unionTag = entry.getKey();
            final var unionStructType = unionTagsReference.get(unionTag);
            final var unionStartingStruct = (Map<String, Object>) startingUnion.get(unionTag);

            return Map.of(unionTag,
                    generateRandomStruct(unionStartingStruct, true, unionStructType.fields, ctx));
        } else {
            final var sortedUnionTagsReference = new ArrayList<>(unionTagsReference.entrySet());

            Collections.sort(sortedUnionTagsReference, (e1, e2) -> e1.getKey().compareTo(e2.getKey()));

            final var randomIndex = ctx.randomGenerator.nextIntWithCeiling(sortedUnionTagsReference.size() - 1);
            final var unionEntry = sortedUnionTagsReference.get(randomIndex);
            final var unionTag = unionEntry.getKey();
            final var unionData = unionEntry.getValue();

            return Map.of(unionTag,
                    generateRandomStruct(null, false, unionData.fields, ctx));
        }
    }
}

File: lib/py/uapi/internal/generation/GenerateRandomUnion.py
from typing import TYPE_CHECKING, cast


if TYPE_CHECKING:
    from uapi.RandomGenerator import RandomGenerator
    from uapi.internal.types.UStruct import UStruct
    from uapi.internal.types.UTypeDeclaration import UTypeDeclaration
    from uapi.internal.generation.GenerateContext import GenerateContext


def generate_random_union(blueprint_value: object, use_blueprint_value: bool,
                          union_tags_reference: dict[str, 'UStruct'], ctx: 'GenerateContext') -> object:
    from uapi.internal.generation.GenerateRandomStruct import generate_random_struct

    if not use_blueprint_value:
        sorted_union_tags_reference = sorted(
            union_tags_reference.items(), key=lambda x: x[0])
        random_index = ctx.random_generator.next_int_with_ceiling(
            len(sorted_union_tags_reference) - 1)
        union_tag, union_data = sorted_union_tags_reference[random_index]
        return {union_tag: generate_random_struct(None, False, union_data.fields, ctx)}
    else:
        starting_union = cast(dict[str, object], blueprint_value)
        union_tag, union_starting_struct = cast(
            tuple[str, dict[str, object]], next(iter(starting_union.items())))
        union_struct_type = union_tags_reference[union_tag]
        return {union_tag: generate_random_struct(union_starting_struct, True, union_struct_type.fields, ctx)}
